chain/txValidator.go

```diff
@@ -264,107 +285,203 @@ func VerifyTransactionWithLedger(txn *transaction.Transaction) error {
 	return nil
 }
 
-type Iterator interface {
-	Iterate(handler func(item *transaction.Transaction) error) error
+type subscription struct {
+	topic      string
+	bucket     uint32
+	subscriber string
+	identifier string
+}
+
+type BlockValidationState struct {
+	txnlist           map[Uint256]struct{}
+	totalAmount       map[Uint160]Fixed64
+	registeredNames   map[string]struct{}
+	nameRegistrants   map[string]struct{}
+	generateIDs       map[string]struct{}
+	subscriptions     map[subscription]struct{}
+	subscriptionCount map[string]int
+}
+
+func NewBlockValidationState() *BlockValidationState {
+	return &BlockValidationState{
+		txnlist:           make(map[Uint256]struct{}, 0),
+		totalAmount:       make(map[Uint160]Fixed64, 0),
+		registeredNames:   make(map[string]struct{}, 0),
+		nameRegistrants:   make(map[string]struct{}, 0),
+		generateIDs:       make(map[string]struct{}, 0),
+		subscriptions:     make(map[subscription]struct{}, 0),
+		subscriptionCount: make(map[string]int, 0),
+	}
 }
 
 // VerifyTransactionWithBlock verifys a transaction with current transaction pool in memory
-func VerifyTransactionWithBlock(ctx context.Context, iterator Iterator, header *block.Header) error {
-	//initial
+func (bvs *BlockValidationState) VerifyTransactionWithBlock(txn *transaction.Transaction, header *block.Header) error {
+	//1.check weather have duplicate transaction.
+	if _, exist := bvs.txnlist[txn.Hash()]; exist {
+		return errors.New("[VerifyTransactionWithBlock], duplicate transaction exist in block.")
+	} else {
+		bvs.txnlist[txn.Hash()] = struct{}{}
+	}
+
+	//3.check issue amount
+	payload, err := transaction.Unpack(txn.UnsignedTx.Payload)
+	if err != nil {
+		return errors.New("[VerifyTransactionWithBlock], payload unpack error.")
+	}
 
-	txnlist := make(map[Uint256]struct{}, 0)
-	registeredNames := make(map[string]struct{}, 0)
-	nameRegistrants := make(map[string]struct{}, 0)
-	generateIDs := make(map[string]struct{}, 0)
+	pg, err := txn.GetProgramHashes()
+	if err != nil {
+		return err
+	}
+	sender := pg[0]
+	var amount Fixed64
+	fee := Fixed64(txn.UnsignedTx.Fee)
 
-	type subscription struct{ topic, subscriber string }
-	subscriptions := make(map[subscription]struct{}, 0)
+	switch txn.UnsignedTx.Payload.Type {
+	case pb.CoinbaseType:
+		coinbase := payload.(*pb.Coinbase)
+		donation, err := DefaultLedger.Store.GetDonation()
+		if err != nil {
+			return err
+		}
+		if header != nil && Fixed64(coinbase.Amount) != GetRewardByHeight(header.UnsignedHeader.Height)+donation.Amount {
```

Comment:
"if header != nil" this judgement can place in front of "coinbase := payload.(*pb.Coinbase)"

Reply:
done